src/nvim/terminal.c

```diff
@@ -288,8 +288,9 @@ void terminal_close(Terminal *term, char *msg)
 
   term->forward_mouse = false;
   term->closed = true;
+  buf_T *buf = handle_get_buffer(term->buf_handle);
```

Comment:
@mhinz  In the `if` block below this, it sets `term->buf_handle = 0`, with the implication that `terminal_close()` may be called again later with that value. `handle_get_buffer()` is a map lookup; if I am reading `build/src/nvim/auto/map.i` correctly, requesting key `0` from it could be a segfault or at least wrong.

Suggestion:

```diff
diff --git a/src/nvim/terminal.c b/src/nvim/terminal.c
index 0a7807d..e616728 100644
--- a/src/nvim/terminal.c
+++ b/src/nvim/terminal.c
@@ -288,8 +288,12 @@ void terminal_close(Terminal *term, char *msg)
 
   term->forward_mouse = false;
   term->closed = true;
+  buf_T *buf = NULL;
+  if (term->buf_handle > 0) {
+    buf = handle_get_buffer(term->buf_handle);
+  }
+
   if (!msg || exiting) {
-    buf_T *buf = handle_get_buffer(term->buf_handle);
     // If no msg was given, this was called by close_buffer(buffer.c).  Or if
     // exiting, we must inform the buffer the terminal no longer exists so that
     // close_buffer() doesn't call this again.
@@ -304,6 +308,10 @@ void terminal_close(Terminal *term, char *msg)
   } else {
     terminal_receive(term, msg, strlen(msg));
   }
+
+  if (buf) {
+    apply_autocmds(EVENT_TERMCLOSE, NULL, NULL, false, buf);
+  }
 }
 
 void terminal_resize(Terminal *term, uint16_t width, uint16_t height)
```